### PR TITLE
PIN-4999 - SECURITY role should not be allowed to delete keys if it is not longer a client member

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala [hunk 1]
```diff
@@ -229,9 +230,19 @@ final case class ClientApiServiceImpl(
...
     logger.info(operationLabel)
 
+    def checkRole(client: PersistentClient): Future[Unit] = for {
+      roles           <- getUserRolesFuture(contexts)
+      requesterUserId <- getUidFutureUUID(contexts)
+      _               <-
+        if (roles.contains(SECURITY_ROLE) && client.users.filter(_ == requesterUserId).isEmpty)
+          Future.failed(KeyOperationNotAllowedOnClient(client.id))
+        else Future.unit
+    } yield ()
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala]: note: even admins cannot perform operations on a client where they are not members, we can skip role check

```suggestion
    def assertUserIsMemberOfClient(client: PersistentClient): Future[Unit] = for {
      requesterUserId <- getUidFutureUUID(contexts)
      _               <-
        if (!client.users.contains(requesterUserId))
          Future.failed(KeyOperationNotAllowedOnClient(client.id))
        else Future.unit
    } yield ()
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala [hunk 2]
```diff
@@ -229,9 +230,19 @@ final case class ClientApiServiceImpl(
...
+
     val result: Future[Unit] = for {
       clientUuid <- clientId.toFutureUUID
       client     <- authorizationManagementService.getClient(clientUuid)
+      _          <- checkRole(client)
       _          <- assertIsClientConsumer(client).toFuture
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala]: let's check first that the requester is the consumer
```suggestion
      _          <- assertIsClientConsumer(client).toFuture
      _          <- checkRole(client)
```

File: src/main/scala/it/pagopa/interop/authorizationprocess/error/AuthorizationProcessErrors.scala [hunk 3]
```diff
@@ -63,4 +63,10 @@ object AuthorizationProcessErrors {
         "0024",
         s"The number of the keys ${size.toString} for the client ${clientId.toString} exceed maximun allowed"
       )
+
+  final case class KeyOperationNotAllowedOnClient(clientId: UUID)
+      extends ComponentError(
+        "0025",
+        s"Role is not allowed to perform operation on the key for the client ${clientId.toString}"
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/error/AuthorizationProcessErrors.scala]:
```suggestion
  final case class UserIsNotMemberOfClient(clientId: UUID, userId: UUID)
      extends ComponentError(
        "0025",
        s"User $userId is not a member of client ${clientId.toString}"
```

**Review by galales — CHANGES_REQUESTED** [review]

File: src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala [hunk 4]
```diff
@@ -230,19 +230,18 @@ final case class ClientApiServiceImpl(
...
+          Future.failed(UserIsNotMemberOfClient(client.id, requesterUserId))
         else Future.unit
     } yield ()
 
     val result: Future[Unit] = for {
       clientUuid <- clientId.toFutureUUID
       client     <- authorizationManagementService.getClient(clientUuid)
-      _          <- checkRole(client)
+      _          <- assertUserIsMemberOfClient(client)
       _          <- assertIsClientConsumer(client).toFuture
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/authorizationprocess/api/impl/ClientApiServiceImpl.scala]: may you move consumer check before member check?
```suggestion
      _          <- assertIsClientConsumer(client).toFuture
      _          <- assertUserIsMemberOfClient(client)
```

**Review by galales — APPROVED** [review]

**Comment by beetlecrunch** [comment]
This is fixed in the new module.

**Pull request closed without merging**